qlib/Logger.qm

```diff
@@ -0,0 +1,2731 @@
+
+        | !Format modifier | !left justify | !minimum width | !maximum width | !comment |
+        | %20c             | false         | 20             | none           | Left pad with spaces if the category name is less than 20 characters long |
+        | %-20c            | true          | 20             | none           | Right pad with spaces if the category name is less than 20 characters long |
+        | %.30c            | NA            | none           | 30             | Truncate from the beginning if the category name is longer than 30 characters |
+        | %20.30c          | false         | 20             | 30             | Left pad with spaces if the category name is shorter than 20 characters. However, if the category name is longer than 30 characters, then truncate from the beginning |
+        | %-20.30c         | true          | 20             | 30             | Right pad with spaces if the category name is shorter than 20 characters. However, if the category name is longer than 30 characters, then truncate from the beginning |
+    */
+    public class LoggerPattern {
+        public {
+            #! format escape character
+            const ESCAPE_CHAR = "%";
+        }
+
+        private:internal {
+            #! pattern
+            string m_pattern;
+        }
+
+        private {
+            #! parsed pattern
+            list m_parsedPattern;  # subclass may provide value e.g. for unit tests
+        }
+
+        #! Creates the object
+        /**
+            @param pattern the logger pattern
+            @throw LOGGER-ERROR when \a pattern is invalid
+        */
+        constructor(string pattern) {
+            # parse pattern
+            setPattern(pattern);
+        }
+
+        #! Sets the new pattern
+        /**
+            @param pattern the logger pattern
+            @throw LOGGER-ERROR when \a pattern is invalid
+        */
+        public setPattern(string pattern) {
+            string s = pattern;
+            list pp = ();
+            while (s != "") {
+                int i = 0;
+                do {
+                    i = bindex(s, ESCAPE_CHAR, i);
+                    if (i < 0) {
+                        i = s.size();
+                    } else {
+                        if (i >= 0 && s[i+1] == ESCAPE_CHAR) {  # %%
+                            splice s, i, 1;
+                            i++;
+                            continue;
+                        }
+                    }
+                    break;
+                } while (True);
+                if (i >= s.size()) {
+                    push pp, s;
+                    s = "";
+                } else {
+                    if (i > 0) {
+                        push pp, substr(s, 0, i);
+                    }
+                    splice s, 0, i+1;
+                    # now the string is token
+%ifdef _DUMMY_TO_PAIR_BRACES_IN_EDITOR
+# but qdx dislikes it when generating docs
+#   {
+%endif
```

Comment:
for what purpose is this hack? Could it be removed?

Reply:
Reason is that curly brackets pairing is broken because of regular expression. Affected are e.g. kate and vim. But qdx is confused as does not recognize conditional section. So correct approach is fixing pattern matching in vim/kate configuration

Reply:
can we use [regex()](https://docs.qore.org/current/lang/html/group__string__functions.html#ga7804f63df9b181e0bacb465eb912feb9) instead? I don't want to have weird `%ifdef` in code only because of editor that was used for development.

also, is this a problem with qore vim plugin only, or is this really broken for all the languages with regular expression?